Avoid requirement to recompile endpoint program when switching policy on/off

ianvernon opened the pull request:
With this change, policy lookup is now performed by default for endpoints in
its PolicyMap. This means that the policy code will be compiled into BPF
programs for endpoints regardless of if there are any rules in the Cilium policy
repository which select said endpoint. Accordingly, remove the corresponding
userspace code which encoded these options into the `lxc_config.h` file
per-endpoint. Convert ingress and egress enforcement state as private fields
within the Endpoint. Policy map contents (and proxy configuration) now are the
sole-source of truth for whether packets are allowed or denied. An empty 
PolicyMap denies all traffic.

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #5325

```release-note
significantly reduce recompilations of endpoint BPF programs after adding any policy rules or removing all policy rules which select the endpoint
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5294)
<!-- Reviewable:end -->

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -471,6 +471,10 @@ type Endpoint struct {
 	// cleaned when this endpoint was first created
 	ctCleaned bool
 
+	ingressPolicy bool
```
`ingressPolicyEnabled`, perhaps?  Also, needs a comment. (Same for `egressPolicy` below).

ianvernon replied:
Will add requisite documentation when I mark as `pending-review` - this was a quick-and-dirty run of the patch to see if it broke any existing tests.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -874,11 +874,7 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
  * Returns true if the packet should pass, false if it should be dropped. */
 static bool check_ip4_fragments(struct iphdr *ip4)
```
The idea behind this function is to return true if either:
* The packet is not an IP fragment, or
* The packet is an IP fragment AND:
  * Policy is disabled on ingress, OR
  * Policy is enabled on ingress, but this traffic would be allowed at L3

And to return false if:
* This packet is an IP fragment, Policy is enabled on ingress, and this traffic would not be allowed at L3

In practice, the policy part of this is a bit naïve and simply checks whether ingress policy enforcement is enabled instead, relying on these `#define`s.

These changes as-is will now always drop any IP fragments that are received. When we run with overlays with kube-proxy, we do not expect to receive any fragments. However, if kube-proxy is not in use, or we run in direct routing mode, it is possible to receive fragments and now we will unconditionally drop them *even if there is no policy in place that might require IP fragments to be reassembled*. (For context, we cannot perform L4+ policy on IP fragments, because IP fragments after the first one do not contain the L4 port and therefore we don't know which port to use to apply L4 policy).

To avoid regressing on this, we would need to have another version of `__policy_can_access()` which operates on fragment packets and skips the L4 lookups (the parts protected by `HAVE_L4_POLICY`) and only allows traffic if it is allowed at L3. Let me know if you'd like me to take a stab at this, but either way this refactoring should be a separate patch from removing `POLICY_INGRESS`/`POLICY_EGRESS`.

ianvernon replied:
@joestringer would be awesome if you could take that effort on. Would you like to do it as a commit within this branch, or an entirely separate patch?

joestringer replied:
Can go either way. I don't expect it to be a huge patch and this PR isn't too big yet.

joestringer replied:
Figured it might be better to put in a separate PR so we can do a full CI run on it and hopefully avoid forcing you to triage any issues that might arise with it: https://github.com/cilium/cilium/pull/5321

Commit:
endpoint: check if DryModeEnabled before sending monitor notification

Errors were appearing during unit tests because no monitor is running for which
notifications could be sent. Thus, check if we are in DryMode accordingly.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: remove POLICY_*GRESS ifdefs

With this change, policy lookup is now performed by default for endpoints in
its PolicyMap. This means that the policy code will be compiled into BPF
programs for endpoints regardless of if there are any rules in the Cilium policy
repository which select said endpoint. Accordingly, remove the corresponding
userspace code which encoded these options into the `lxc_config.h` file
per-endpoint. Convert ingress and egress enforcement state as private fields
within the Endpoint.

Signed-off by: Ian Vernon <ian@cilium.io>

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -476,6 +484,48 @@ type Endpoint struct {
+}
+
+// GetEgressPolicyEnabledUnlocked returns whether egress policy enforcement
+// is enabled for endpoint or not. The endpoint's mutex must be held.
```
Same here

ianvernon replied:
Fixed.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -476,6 +484,48 @@ type Endpoint struct {
+// GetIngressPolicyEnabledUnlocked returns whether ingress policy
+// enforcement is enabled for endpoint or not. The endpoint's mutex must be
+// held.
+func (e *Endpoint) GetIngressPolicyEnabledUnlocked() bool {
```
Is this supposed to be `GetIngressPolicyEnabledLocked()`?

ianvernon replied:
You’re right, I was using Unlocked to refer to no locking in the function itself, but that’s inconsistent with the rest of the code and is confusing. I’ll change it.

ianvernon replied:
Fixed.
